File: ghc-4335c07/nofib/real/gamteb/inLML/photoElec.m
-- 
--      Patricia Fasel
--      Los Alamos National Laboratory
--      1990 August
--
module

#include "gamtebType.t"
#include "consts.t"
#include "utils.t"

export photoElec;


rec
	-- photoElec :: Particle -> Prob -> (Particle, Float, Bool)

    photoElec (Part pos dir w e eIndx cell seed)
		(Pr pComp pPair pPhot pTot) =
	let (seed', r2) = genRand seed in
	let (r1, r3) = genRand r2 in
	let w' = w *. (1.0-.pPhot/.pTot) in
	let absorb = w-.w' in
	let fcell = itof cell in
	if (w' > wcp2)
	    then 
		(Part pos dir w' e eIndx cell seed', absorb, false)
	    else	-- terminate particle because of weight cutoff 
		if ((w'*.fcell) < (r1*.wcp1))
		    then (Part pos dir w' e eIndx cell seed', absorb, true)
		    else (Part pos dir (wcp1/.fcell) e eIndx cell seed', absorb, false)

end

File: ghc-4335c07/nofib/real/gamteb/PhotoElec.hs
-- 
--      Patricia Fasel
--      Los Alamos National Laboratory
--      1990 August
--
module PhotoElec (photoElec) where

import GamtebType
import Consts
import Utils

photoElec :: Particle -> Probability -> (Particle, Value, Bool)
photoElec p prob =
	if (w' > wgtCut2)
	  then (Part pos dir w' e eIndx cell seed', absorb, False)
	  else	-- terminate particle because of weight cutoff 
	    if ((w' * fromIntegral cell) < (r1 * wgtCut))
	      then (Part pos dir w' e eIndx cell seed', absorb, True)
	      else (Part pos dir w'' e eIndx cell seed', absorb, False)
	where
	    (Part pos dir w e eIndx cell seed) = p
	    (pComp,pPair,pPhot,pTot) = prob
	    (seed', r2) = genRand seed
	    (r1, r3) = genRand r2
	    w' = w * (1 - pPhot/pTot)
	    w'' = wgtCut/fromIntegral cell
	    absorb = w - w'
